tuftool/src/clone.rs

```diff
@@ -0,0 +1,144 @@
+            std::process::exit(1);
+        };
+
+        // Structopt won't allow `targets_base_url` to be None when it is required.  We require the
+        // user to supply `targets_base_url` in the case they actually plan to download targets.
+        // When downloading metadata, we don't ever need to access the targets URL, so we use a
+        // fake URL to satisfy the library.
+        let targets_base_url = self
+            .targets_base_url
+            .as_ref()
+            .unwrap_or(&Url::parse(UNUSED_URL).context(error::UrlParse {
+                url: UNUSED_URL.to_owned(),
+            })?)
+            .clone();
+
+        // Load repository
+        let expiration_enforcement = if self.allow_expired_repo {
+            expired_repo_warning(&self.metadata_dir);
+            ExpirationEnforcement::Unsafe
+        } else {
+            ExpirationEnforcement::Safe
+        };
+        let repository = RepositoryLoader::new(
+            File::open(&root_path).context(error::OpenRoot { path: &root_path })?,
+            self.metadata_base_url.clone(),
+            targets_base_url,
+        )
+        .expiration_enforcement(expiration_enforcement)
+        .load()
+        .context(error::RepoLoad)?;
+
+        // Clone the repository, downloading none, all, or a subset of targets
+        let clone_result = if self.metadata_only {
+            println!("Cloning repository metadata to {:?}", self.metadata_dir);
+            repository.cache_metadata(&self.metadata_dir, true)
+        } else {
+            // Similar to `targets_base_url, structopt's guard rails won't let us have a
+            // `targets_dir` that is None when the argument is required.  We only require the user
+            // to supply a targets directory if they actually plan on downloading targets.
+            let targets_dir = self.targets_dir.as_ref().expect(
+                "Developer error: `targets_dir` is required unless downloading metadata only",
+            );
+
+            if self.target_names.is_empty() {
+                println!(
+                    "Cloning repository:\n\tmetadata location: {:?}\n\ttargets location: {:?}",
+                    self.metadata_dir, targets_dir
+                );
+                // The None turbofish is required to satisfy the compiler, sorry.
+                repository.cache(&self.metadata_dir, &targets_dir, None::<&[&str]>, true)
+            } else {
+                println!(
+                    "Cloning repository with a subset of targets:\n\tmetadata location:{:?}\n\ttargets location: {:?}",
+                    self.metadata_dir, targets_dir
+                );
+                repository.cache(
+                    &self.metadata_dir,
+                    &targets_dir,
+                    Some(self.target_names.as_slice()),
+                    true,
+                )
+            }
+        };
+
+        clone_result.context(error::CloneRepository)
```

Comment:
nit: I think this context, long after the storage of the Result, is throwing off the flow for readers more than it's helping; I'd vote for putting the context on each call, with a `?` so points of failure are more obvious, and letting the function end "normally."

Reply:
Makes sense - I went back and forth on it before going this way. :)